### PR TITLE
Eagle3 DP attention for Qwen3 MoE

### PR DESCRIPTION
**Pull request by qhsc**
## Motivation

Add eagle3 dp attention for Qwen3 MoE model for large scale EP deployment.

## Benchmarking and Profiling
We test DP attention for Eagle3 in non-PD and PD scenarios. The results are all good.

### Single node of 8xH100 using Eagle and DP attention
engine command
```bash
python3 -m sglang.launch_server --model-path ~/models/Qwen3-235B-A22B --trust-remote-code --disable-radix-cache --tp-size 8 --dp-size 2 --mem-fraction-static 0.85 --context-length 32768 --enable-dp-attention --enable-dp-lm-head --moe-dense-tp-size 1 --speculative-algorithm EAGLE3 --speculative-num-steps 3 --speculative-eagle-topk 3 --speculative-num-draft-tokens 8 --speculative-draft-model-path ~/models/lmsys/Qwen3-235B-A22B-EAGLE3 --attention-backend fa3
```
test command and result
```
$ python bench_sglang.py --port 30000 --num-questions 200
Accuracy: 0.970
Invalid: 0.000
Latency: 48.009 s
Output throughput: 565.025 token/s
```

### 3 nodes of 8xH100 using 1P1D PD with Eagel and DP attention
engine commands
```bash
# prefill, one node 8xH100
python3 -m sglang.launch_server --disaggregation-mode prefill --model-path ~/models/Qwen3-235B-A22B --host 0.0.0.0 --trust-remote-code --disable-radix-cache --tp-size 8 --mem-fraction-static 0.85 --max-running-requests 64 --cuda-graph-max-bs 32 --context-length 32768 --speculative-algorithm EAGLE3 --speculative-num-steps 3 --speculative-eagle-topk 3 --speculative-num-draft-tokens 8 --speculative-draft-model-path ~/models/lmsys/Qwen3-235B-A22B-EAGLE3 --attention-backend fa3

# decode, two node 8xH100. Note: we use a fp8 backbone model
 python3 -m sglang.launch_server --disaggregation-mode decode --model-path ~/models/Qwen3-235B-A22B-FP8 --host 0.0.0.0 --trust-remote-code --nnodes 2 --node-rank $RANK --disable-radix-cache --tp-size 16 --dp-size 2 --ep-size 16 --mem-fraction-static 0.85 --max-running-requests 128 --cuda-graph-max-bs 32 --context-length 32768 --speculative-algorithm EAGLE3 --speculative-num-steps 3 --speculative-eagle-topk 3 --speculative-num-draft-tokens 8 --speculative-draft-model-path ~/models/lmsys/Qwen3-235B-A22B-EAGLE3 --enable-dp-attention --enable-dp-lm-head --moe-dense-tp-size 1 --moe-a2a-backend deepep --deepep-mode low_latency --eplb-algorithm deepseek --ep-dispatch-algorithm static --dist-init-addr ${ADDR} --load-balance-method round_robin --prefill-round-robin-balance

```

test command and result
```
$ python bench_sglang.py --port 8192 --num-questions 200
Accuracy: 0.965
Invalid: 0.000
Latency: 21.881 s
Output throughput: 1253.305 token/s
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_worker.py [hunk 1]
```diff
@@ -117,7 +118,12 @@ def __init__(
             self.hot_token_id = None
 
         # Init draft worker
-        with empty_context():
+        tp_group = None
+        ctx = empty_context()
```
Comment (mickqian) [inline python/sglang/srt/speculative/eagle_worker.py]: nit: could you move these to else branch?

Comment (qhsc) [reply python/sglang/srt/speculative/eagle_worker.py]: Thanks~ I'll do it.

File: python/sglang/srt/models/llama_eagle3.py [hunk 2]
```diff
@@ -212,6 +217,12 @@ def __init__(
                 prefix=add_prefix("lm_head", prefix),
             )
 
+        config_ = copy.deepcopy(config)
+        config_.vocab_size = (
+            config_.draft_vocab_size
+        )  # draft logits processor has it's own vocab size
+        self.logits_processor = LogitsProcessor(config_)
+
         self.logits_processor = LogitsProcessor(config)
```
Comment (yizhang2077) [inline python/sglang/srt/models/llama_eagle3.py]: maybe you need remove this duplicate `logins_processor`?

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -828,7 +824,7 @@ def _generate_cuda_graph_batch_sizes(self):
             capture_bs = (
                 list(range(1, 9, 1))
                 + list(range(10, 33, 2))
-                + list(range(40, 64, 4))
+                + list(range(40, 65, 4))
```
Comment (yizhang2077) [inline python/sglang/srt/server_args.py:827]: why do we need change here?

File: python/sglang/srt/layers/communicator.py [hunk 4]
```diff
@@ -216,6 +216,27 @@ def __init__(
...
+            hidden_states, residual, forward_batch
+        )
+        if captured_last_layer_outputs is not None:
+            gathered_last_layer_output = self._communicate_simple_fn(
+                hidden_states=residual,
+                forward_batch=forward_batch,
+                context=self._context,
+            )
+            if gathered_last_layer_output is residual:
+                gathered_last_layer_output = residual.clone()
```
Comment (yizhang2077) [inline python/sglang/srt/layers/communicator.py:237]: why do we need clone here? cc @fzyzcjy may you take a look for this part?

**Comment by yizhang2077** [comment]
I think we need to add some unit tests, including different tp/dp/ep (deepep) config, and ensure accuracy and accept length. ref: `TestEAGLE3Engine`

File: python/sglang/srt/server_args.py (hunk 3 above)
Comment (qhsc) [reply python/sglang/srt/server_args.py:827]: It will skip bs 64, just like this [... 56, 60, 72, 80, 88, ...]. The gap between 60 and 72 seems like a mistake.

Comment (yizhang2077) [reply python/sglang/srt/server_args.py:827]: make sense

File: python/sglang/srt/layers/communicator.py (hunk 4 above)
Comment (yizhang2077) [reply python/sglang/srt/layers/communicator.py:237]: The main usage for this function is to transfer hidden states from scatter to tp full when using deepep?

Comment (qhsc) [reply python/sglang/srt/layers/communicator.py:237]: It will be modified by some following operation. I'll check it again.

File: python/sglang/srt/models/llama_eagle3.py (hunk 2 above)
Comment (qhsc) [reply python/sglang/srt/models/llama_eagle3.py]: Yeah, a mistake...

File: python/sglang/srt/layers/communicator.py (hunk 4 above)
Comment (qhsc) [reply python/sglang/srt/layers/communicator.py:237]: > The main usage for this function is to transfer hidden states from scatter to tp full when using deepep?

yes.

Comment (qhsc) [reply python/sglang/srt/layers/communicator.py:237]: > It will be modified by some following operation. I'll check it again.

The `fused_add_rmsnorm` function in RMSNorm will do this `residual[i] += input[i]`.

File: python/sglang/srt/models/llama_eagle3.py [hunk 5]
```diff
@@ -161,6 +162,10 @@ def forward(
         if hidden_states.shape[-1] != embeds.shape[-1]:
             hidden_states = self.fc(hidden_states)
 
+        # idle batch
+        if hidden_states.shape[0] == 0:
+            return hidden_states, [hidden_states]
```
Comment (yizhang2077) [inline python/sglang/srt/models/llama_eagle3.py:167]: I think we need add dp attention support for llama eagle3 model?

Comment (qhsc) [reply python/sglang/srt/models/llama_eagle3.py:167]: It's already supported, sine we initialize eagle worker with `ctx = draft_tp_context(get_attention_tp_group())`. Eagle model will work with attention tp group.

**Comment by qhsc** [comment]
> I think we need to add some unit tests, including different tp/dp/ep (deepep) config, and ensure accuracy and accept length. ref: `TestEAGLE3Engine`

Good idea, let me check this.

File: python/sglang/srt/models/llama_eagle3.py (hunk 5 above)
Comment (yizhang2077) [reply python/sglang/srt/models/llama_eagle3.py:167]: I think we still need add communicator adaption to do gather/reduce_scatter/scatter?

Comment (qhsc) [reply python/sglang/srt/models/llama_eagle3.py:167]: let me check it.

Comment (qhsc) [reply python/sglang/srt/models/llama_eagle3.py:167]: > I think we still need add communicator adaption to do gather/reduce_scatter/scatter?

I’ve double-checked — Eagle3 model doesn’t need an extra communicator, since the captured hidden states have already been handled with communicator operations inside the backbone model.
By the way, I’ve also added a test for DP2 TP8, and it passed. In the description section, I’ve included the EP/PD test results as well — please refer to them. Thanks！

Comment (yizhang2077) [reply python/sglang/srt/models/llama_eagle3.py:167]: > It's already supported, sine we initialize eagle worker with `ctx = draft_tp_context(get_attention_tp_group())`. Eagle model will work with attention tp group.

I see, by using `draft_tp_context` each dp attention group has an individual draft model so it do not needs communicator.

File: test/srt/test_eagle_infer_a.py [hunk 6]
```diff
@@ -396,5 +403,112 @@ def setUpClass(cls):
         cls.accept_len_threshold = 1.85
 
 
+class TestEAGLE3EngineDPAttention(CustomTestCase):
```
Comment (yizhang2077) [inline test/srt/test_eagle_infer_a.py]: You may need move this test into individual file since it needs more gpu for ci (this file only needs 1 gpu for each test).  And you may also need add the test name into `run_suite.py` in  `per-commit-8-gpu-h200-deepep` for triggering ci.

File: python/sglang/srt/models/llama_eagle3.py (hunk 5 above)
Comment (qhsc) [reply python/sglang/srt/models/llama_eagle3.py:167]: > > It's already supported, sine we initialize eagle worker with `ctx = draft_tp_context(get_attention_tp_group())`. Eagle model will work with attention tp group.
> 
> I see, by using `draft_tp_context` each dp attention group has an individual draft model so it do not needs communicator.

exactly. for eagle dense layer，using attn tp  group seems like not bad.

Comment (yizhang2077) [reply python/sglang/srt/models/llama_eagle3.py:167]: make sense

File: python/sglang/test/test_utils.py [hunk 7]
```diff
@@ -84,6 +84,8 @@
 DEFAULT_EAGLE_TARGET_MODEL_FOR_TEST = "meta-llama/Llama-2-7b-chat-hf"
 DEFAULT_EAGLE_DRAFT_MODEL_FOR_TEST = "lmsys/sglang-EAGLE-llama2-chat-7B"
 DEFAULT_EAGLE_TARGET_MODEL_FOR_TEST_EAGLE3 = "meta-llama/Llama-3.1-8B-Instruct"
+DEFAULT_EAGLE_DP_ATTENTION_TARGET_MODEL_FOR_TEST = "Qwen/Qwen3-235B-A22B"
```
Comment (yizhang2077) [inline python/sglang/test/test_utils.py]: This model may be heavy for running ci. cc @ispobock

File: test/srt/test_eagle_infer_a.py (hunk 6 above)
Comment (qhsc) [reply test/srt/test_eagle_infer_a.py]: Thanks, I'll fix it.

File: test/srt/test_eagle_infer_a.py [hunk 8]
```diff
@@ -396,5 +403,112 @@ def setUpClass(cls):
...
+
+        print(f"{avg_spec_accept_length=}")
+
+        if is_in_ci():
+            write_github_step_summary(
+                f"### test_gsm8k (EAGLE3 DP Attention)\n"
+                f'{metrics["accuracy"]=:.3f}\n'
+                f"{avg_spec_accept_length=:.2f}\n"
+            )
+            self.assertGreater(metrics["accuracy"], 0.90)
```
Comment (yizhang2077) [inline test/srt/test_eagle_infer_a.py]: maybe we need a higher value? (like 0.92 or 0.93?)

Comment (qhsc) [reply test/srt/test_eagle_infer_a.py]: ok, I'll change to 0.93

File: python/sglang/test/test_utils.py (hunk 7 above)
Comment (qhsc) [reply python/sglang/test/test_utils.py]: How about Qwen/Qwen3-30B-A3B?

Comment (yizhang2077) [reply python/sglang/test/test_utils.py]: it is ok

File: python/sglang/srt/models/qwen2_moe.py [hunk 9]
```diff
@@ -830,13 +843,15 @@ def set_eagle3_layers_to_capture(self, layer_ids: Optional[List[int]] = None):
         self.capture_aux_hidden_states = True
         if layer_ids is None:
             num_layers = self.config.num_hidden_layers
-            self.model.layers_to_capture = [
-                2,
-                num_layers // 2,
-                num_layers - 3,
-            ]  # Specific layers for EAGLE3 support
+            self.model.set_eagle3_layers_to_capture(
```
Comment (sleepcoo) [inline python/sglang/srt/models/qwen2_moe.py:846]: The duplicate names are a bit confusing.

**Review by yizhang2077 — APPROVED** [review]
LGTM, once ci is passed it can be merged cc @ispobock

**Review by sleepcoo — APPROVED** [review]

File: test/srt/test_eagle_dp_attention.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class TestEAGLE3EngineDPAttention(CustomTestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.model = DEFAULT_EAGLE_DP_ATTENTION_TARGET_MODEL_FOR_TEST
+        cls.base_url = DEFAULT_URL_FOR_TEST
+        other_args = [
+            "--trust-remote-code",
+            "--disable-radix-cache",
```
Comment (ispobock) [inline test/srt/test_eagle_dp_attention.py]: can we open radix cache in test?

File: test/srt/run_suite.py [hunk 11]
```diff
@@ -184,6 +184,7 @@ class TestFile:
     ],
     "per-commit-8-gpu-h200-deepep": [
         TestFile("ep/test_deepep_large.py", 338),
+        TestFile("test_eagle_dp_attention.py", 300),
```
Comment (ispobock) [inline test/srt/run_suite.py]: for 30B model, we don't need to put it in 8 gpu h200 runner

Comment (qhsc) [reply test/srt/run_suite.py]: I'll fix it

File: test/srt/test_eagle_dp_attention.py (hunk 10 above)
Comment (qhsc) [reply test/srt/test_eagle_dp_attention.py]: sure.

File: test/srt/run_suite.py (hunk 11 above)
Comment (qhsc) [reply test/srt/run_suite.py]: > for 30B model, we don't need to put it in 8 gpu h200 runner

I've updated it. I also changed the score threshold of gsm8k from `0.93` to `0.91`, this is more reasonable for 30B-A3B model according to Qwen3 tech report.

File: test/srt/test_eagle_dp_attention.py
Comment (ch-wan) [inline test/srt/test_eagle_dp_attention.py:1]: I think it's covered in CI. https://github.com/sgl-project/sglang/blob/9fc3e8aac7422826a7fad3477ddafe74e03e6420/test/srt/test_dp_attention.py#L61

File: python/sglang/srt/speculative/eagle_worker.py
Comment (ch-wan) [inline python/sglang/srt/speculative/eagle_worker.py:1]: dp attn + mtp was supported. why do we need to change this file?

File: python/sglang/srt/models/qwen2_moe.py [hunk 12]
```diff
@@ -553,6 +559,11 @@ def __init__(
         # For EAGLE3 support
         self.layers_to_capture = []
 
+    def set_eagle3_layers_to_capture(self, layers_to_capture: List[int]):
```
Comment (ch-wan) [inline python/sglang/srt/models/qwen2_moe.py:562]: the same function has been defined in this module.

File: python/sglang/srt/speculative/eagle_worker.py
Comment (qhsc) [reply python/sglang/srt/speculative/eagle_worker.py:1]: Yeah, MTP DP attention is supported for attention tp group size equals 1. But eagle3 model for attention tp group not equls 1 will fail, for example, `--tp-size 8, --dp-size=2` is not a valid config.

File: test/srt/test_eagle_dp_attention.py
Comment (qhsc) [reply test/srt/test_eagle_dp_attention.py:1]: MTP and eagle3 model using different model arch, inclding attention and MLP module. We should also test them.

File: python/sglang/srt/models/qwen2_moe.py (hunk 12 above)
Comment (qhsc) [reply python/sglang/srt/models/qwen2_moe.py:562]: I don't understand that. Is that means both `Qwen2MoeForCausalLM` and `Qwen2MoeModel` have the function `set_eagle3_layers_to_capture` is not allowed?

File: test/srt/run_suite.py (hunk 11 above)
Comment (qhsc) [reply test/srt/run_suite.py]: @ispobock Hi, please check the CI results. The failures don’t seem related to this PR.
